src/Sentry/Protocol/Metrics/Metric.cs

```diff
@@ -0,0 +1,126 @@
+using Sentry.Extensibility;
+using Sentry.Internal.Extensions;
+using ISentrySerializable = Sentry.Protocol.Envelopes.ISerializable;
+
+namespace Sentry.Protocol.Metrics;
+
+internal abstract class Metric : IJsonSerializable, ISentrySerializable
+{
+    protected Metric() : this(string.Empty)
+    {
+    }
+
+    protected Metric(string key, MeasurementUnit? unit = null, IDictionary<string, string>? tags = null, DateTime? timestamp = null)
+    {
+        Key = key;
+        Unit = unit;
+        Tags = tags ?? new Dictionary<string, string>();
+        Timestamp = timestamp ?? DateTime.UtcNow;
+    }
+
+    public SentryId EventId { get; private set; } = SentryId.Create();
+
+    public string Key { get; private set; }
+
+    public DateTime Timestamp { get; private set; }
+
+    public MeasurementUnit? Unit { get; private set; }
+
+    public IDictionary<string, string> Tags { get; private set; }
+
+    public abstract void Add(double value);
+
+    protected abstract void WriteValues(Utf8JsonWriter writer, IDiagnosticLogger? logger);
+
+    public void WriteTo(Utf8JsonWriter writer, IDiagnosticLogger? logger)
+    {
+        writer.WriteStartObject();
+        writer.WriteString("type", GetType().Name);
+        writer.WriteSerializable("event_id", EventId, logger);
+        writer.WriteString("name", Key);
+        writer.WriteString("timestamp", Timestamp);
+        if (Unit.HasValue)
+        {
+            writer.WriteStringIfNotWhiteSpace("unit", Unit.ToString());
+        }
+        writer.WriteStringDictionaryIfNotEmpty("tags", Tags!);
```

Comment:
Looks like Tags is never null. Do we need `!` here?

Reply:
If I remove it, we get a compiler error due to the signature of the ExtensionMethod being called:
> Argument of type 'System.Collections.Generic.IDictionary<string,string>' cannot be used for parameter 'dic' of type 'System.Collections.Generic.IEnumerable<System.Collections.Generic.KeyValuePair<string,string?>>?' in 'Sentry.Internal.Extensions.JsonExtensions.WriteStringDictionaryIfNotEmpty' because of differences in the nullability of reference types

I can't think of an elegant way aroud this. I could disable the warning with a pragma and include a comment?

Reply:
Yeah, it's a bit weird. Another way to make the warning go away would be an explicit cast:
![image](https://github.com/getsentry/sentry-dotnet/assets/728212/b6c37eb9-f771-4e71-b55a-1df4b2c16f0c)

You can see the first call to `WriteStringDictionary` has no warning... where the second does. It's a lot more verbose than a Null forgiving operator but 🤷🏻